vumi_message_store/memory_backend_manager.py

```diff
@@ -0,0 +1,371 @@
+        else:
+            self.call_decorator = inlineCallbacks
+        self.store_versions = store_versions or {}
+        # Wrap the methods that need to handle async stuff.
+        self.load = self.call_decorator(self._internal_load)
+        self.store = self.call_decorator(self._internal_store)
+
+    def proxy(self, modelcls):
+        """
+        Return a real proxy using the fake manager.
+        """
+        return ModelProxy(self, modelcls)
+
+    def bucket_name(self, modelcls_or_obj):
+        return modelcls_or_obj.bucket
+
+    def bucket_for_modelcls(self, modelcls):
+        """
+        Return a fake bucket object for the given model class.
+        """
+        bucket_name = self.bucket_name(modelcls)
+        return FakeRiakBucket(self._state, bucket_name)
+
+    def riak_object(self, modelcls, key):
+        """
+        Create and return a fake RiakObject.
+        """
+        bucket_name = self.bucket_name(modelcls)
+        riak_object = FakeRiakObject(self._state, bucket_name, key)
+        riak_object.set_content_type("application/json")
+        riak_object.set_data({'$VERSION': modelcls.VERSION})
+        return riak_object
+
+    def _internal_store(self, modelobj):
+        """
+        This method gets wrapped in a sync or async decorator in __init__().
+        """
+        riak_object = modelobj._riak_object
+        modelcls = type(modelobj)
+        model_name = "%s.%s" % (modelcls.__module__, modelcls.__name__)
+        store_version = self.store_versions.get(model_name, modelcls.VERSION)
+        # Run reverse migrators until we have the correct version of the data.
+        data_version = riak_object.get_data().get('$VERSION', None)
+        while data_version != store_version:
+            migrator = modelcls.MIGRATOR(
+                modelcls, self, data_version, reverse=True)
+            riak_object = migrator(riak_object).get_riak_object()
+            data_version = riak_object.get_data().get('$VERSION', None)
+        yield riak_object.store()
+        returnValue(modelobj)
+
+    def _internal_load(self, modelcls, key, result=None):
+        """
+        This method gets wrapped in a sync or async decorator in __init__().
+        """
+        assert result is None
+        riak_object = self.riak_object(modelcls, key)
+        yield riak_object.reload()
+        was_migrated = False
+
+        # Run migrators until we have the correct version of the data.
+        while riak_object.get_data() is not None:
+            data_version = riak_object.get_data().get('$VERSION', None)
+            if data_version == modelcls.VERSION:
+                obj = modelcls(self, key, _riak_object=riak_object)
+                obj.was_migrated = was_migrated
+                returnValue(obj)
+            migrator = modelcls.MIGRATOR(modelcls, self, data_version)
+            riak_object = migrator(riak_object).get_riak_object()
+            was_migrated = True
```

Comment:
If the object we're loading doesn't exist. This could probably be rewritten as `if riak_object.get_data() is None: returnValue(None)` followed by a `while True:` loop, but this code was copied from our existing managers which do it this way.